File: sdk/python/tempmail_sdk/providers/disposablemail_com.py
"""
disposablemail.com 临时邮箱渠道（METRONET s.r.o. 后端）

与 fakemail.net / minuteinbox.com 共享相同的 PHP API 结构。
流程：GET / 获取 session cookie + CSRF token → GET /index/index?csrf_token={csrf} 创建邮箱
      GET /index/refresh 获取邮件列表 → GET /email/id/{id} 获取邮件 HTML 正文
token 序列化保存 CSRF token 与 session cookie 串，供 get_emails 复用。
"""

import base64
import json
import re
from typing import Dict, List, Tuple

import requests

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "disposablemail-com"
BASE_URL = "https://www.disposablemail.com"
_TOK_PREFIX = "dmc1:"

# 从 HTML 中提取 CSRF token（格式: CSRF="xxx"）
_CSRF_RE = re.compile(r'CSRF="([^"]+)"')

_UA = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"
)


def _browser_headers() -> Dict[str, str]:
    return {
        "User-Agent": _UA,
        "Accept": (
            "text/html,application/xhtml+xml,application/xml;q=0.9,"
            "image/avif,image/webp,image/apng,*/*;q=0.8"
        ),
        "Accept-Language": "en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7",
    }


def _ajax_headers(cookie_hdr: str) -> Dict[str, str]:
    headers = {
        "User-Agent": _UA,
        "Accept": "application/json, text/javascript, */*; q=0.01",
        "Accept-Language": "en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7",
        "X-Requested-With": "XMLHttpRequest",
        "Referer": f"{BASE_URL}/",
    }
    if cookie_hdr:
        headers["Cookie"] = cookie_hdr
    return headers


def _cookie_header(prev: str, resp: requests.Response) -> str:
    """合并已有 cookie 串与响应新 cookie（按键排序）"""
    d: Dict[str, str] = {}
    for part in (prev or "").split(";"):
        part = part.strip()
        if "=" in part:
            k, _, v = part.partition("=")
            k = k.strip()
            if k:
                d[k] = v.strip()
    d.update(resp.cookies.get_dict())
    return "; ".join(f"{k}={d[k]}" for k in sorted(d.keys()))


def _encode_token(csrf: str, cookie_hdr: str) -> str:
    raw = json.dumps({"t": csrf, "c": cookie_hdr}, separators=(",", ":")).encode("utf-8")
    return _TOK_PREFIX + base64.b64encode(raw).decode("ascii")


def _decode_token(token: str) -> Tuple[str, str]:
    if not token.startswith(_TOK_PREFIX):
        raise ValueError("disposablemail-com: 无效的 token")
    try:
        data = base64.b64decode(token[len(_TOK_PREFIX):])
        o = json.loads(data.decode("utf-8"))
    except (ValueError, UnicodeDecodeError) as e:
        raise ValueError("disposablemail-com: 无效的 token") from e
    csrf = str(o.get("t") or "").strip()
    cookie_hdr = str(o.get("c") or "").strip()
    if not csrf:
        raise ValueError("disposablemail-com: token 中缺少 CSRF")
    return csrf, cookie_hdr


def generate_email(channel: str = CHANNEL) -> EmailInfo:
    """
    创建 disposablemail.com 临时邮箱

    1. GET / 获取 session cookie 和 CSRF token（正则匹配 CSRF="xxx"）
    2. GET /index/index?csrf_token={csrf} 创建邮箱，返回 {"email":"user@domain.com"}
    token 序列化保存 CSRF 与 session cookie。
    """
    resp = tm_http.get(f"{BASE_URL}/", headers=_browser_headers())
    resp.raise_for_status()
    cookie_hdr = _cookie_header("", resp)

    m = _CSRF_RE.search(resp.text)
    if not m or not m.group(1):
        raise RuntimeError("disposablemail-com: 未能从首页提取 CSRF token")
    csrf = m.group(1)

    resp2 = tm_http.get(
        f"{BASE_URL}/index/index",
        headers=_ajax_headers(cookie_hdr),
        params={"csrf_token": csrf},
    )
    resp2.raise_for_status()
    cookie_hdr = _cookie_header(cookie_hdr, resp2)

    data = resp2.json()
    if not isinstance(data, dict):
        raise RuntimeError("disposablemail-com: 创建邮箱响应格式异常")
    email = str(data.get("email") or "").strip()
    if not email or "@" not in email:
        raise RuntimeError(f"disposablemail-com: 获取到的邮箱地址无效: {email!r}")

    return EmailInfo(channel=channel, email=email, _token=_encode_token(csrf, cookie_hdr))


def get_emails(email: str, token: str) -> List[Email]:
    """
    获取 disposablemail.com 邮件列表

    1. GET /index/refresh 获取邮件列表（JSON 数组，空收件箱返回数字 0）
    2. 对每封邮件 GET /email/id/{id} 获取 HTML 正文
    字段（捷克语）: predmet=subject, od=from, id=邮件ID, kdy=when, precteno=read status
    """
    email = (email or "").strip()
    if not email:
        raise ValueError("disposablemail-com: 邮箱地址为空")
    _csrf, cookie_hdr = _decode_token((token or "").strip())

    resp = tm_http.get(f"{BASE_URL}/index/refresh", headers=_ajax_headers(cookie_hdr))
    resp.raise_for_status()

    trimmed = resp.text.strip()
    if trimmed in ("0", "", "[]"):
        return []

    mail_list = resp.json()
    if not isinstance(mail_list, list) or not mail_list:
        return []

    out: List[Email] = []
    for item in mail_list:
        if not isinstance(item, dict):
            continue
        mail_id = str(item.get("id") or "")
        raw = {
            "id": mail_id,
            "from": item.get("od", ""),
            "to": email,
            "subject": item.get("predmet", ""),
            "html": _fetch_body(cookie_hdr, mail_id),
            "date": item.get("kdy", ""),
            "isRead": item.get("precteno") == "precteno",
        }
        out.append(normalize_email(raw, email))

    return out


def _fetch_body(cookie_hdr: str, mail_id: str) -> str:
    """获取单封邮件的 HTML 正文（GET /email/id/{id}），失败返回空串"""
    if not mail_id:
        return ""
    try:
        resp = tm_http.get(f"{BASE_URL}/email/id/{mail_id}", headers=_ajax_headers(cookie_hdr))
    except OSError:
        return ""
    if resp.status_code < 200 or resp.status_code >= 300:
        return ""
    return resp.text

File: sdk/npm/src/providers/disposablemail-com.ts
/**
 * disposablemail.com 临时邮箱渠道（METRONET s.r.o. 后端）
 *
 * 与 minuteinbox.com 共享相同的 PHP API 结构。
 * 流程：
 *   1. GET / 获取 session cookie + CSRF token（正则 CSRF="xxx"）
 *   2. GET /index/index?csrf_token={csrf} 创建邮箱，返回 {"email":"user@domain.com"}
 *   3. GET /index/refresh 获取邮件列表（空收件箱返回数字 0）
 *   4. GET /email/id/{id} 获取邮件 HTML 正文
 *
 * Go 端 token 仅存 CSRF（session 由全局 cookie jar 维护）；本端无 cookie jar，
 * 故 token 以 base64 JSON 同时存储 {csrf, cookie}，getEmails 时手动携带 Cookie。
 * 字段说明（捷克语）: predmet=subject, od=from, id=邮件ID, kdy=when, precteno=read status
 */
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "disposablemail-com";
const BASE_URL = "https://www.disposablemail.com";

/** 从 HTML 中提取 CSRF token（格式: CSRF="xxx"） */
const CSRF_RE = /CSRF="([^"]+)"/;

const USER_AGENT =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0";

interface DisposablemailRow {
  predmet?: string;
  od?: string;
  id?: string | number;
  kdy?: string;
  precteno?: string;
}

function browserHeaders(): Record<string, string> {
  return {
    "User-Agent": USER_AGENT,
    Accept:
      "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
    "Accept-Language": "en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7",
  };
}

function ajaxHeaders(cookie: string): Record<string, string> {
  return {
    "User-Agent": USER_AGENT,
    Accept: "application/json, text/javascript, */*; q=0.01",
    "Accept-Language": "en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7",
    "X-Requested-With": "XMLHttpRequest",
    Referer: `${BASE_URL}/`,
    ...(cookie ? { Cookie: cookie } : {}),
  };
}

/** 从 Set-Cookie 响应头提取 cookie 行（Node 下 Headers 可能有 getSetCookie） */
function setCookieLines(headers: Headers): string[] {
  const h = headers as Headers & { getSetCookie?: () => string[] };
  if (typeof h.getSetCookie === "function") {
    return h.getSetCookie();
  }
  const one = headers.get("set-cookie");
  return one ? [one] : [];
}

/** 将 Set-Cookie 合并进既有 Cookie 请求头值 */
function mergeCookieHeader(prev: string, headers: Headers): string {
  const map = new Map<string, string>();
  if (prev) {
    for (const part of prev.split(";")) {
      const eq = part.indexOf("=");
      if (eq <= 0) continue;
      map.set(part.slice(0, eq).trim(), part.slice(eq + 1).trim());
    }
  }
  for (const line of setCookieLines(headers)) {
    const first = line.split(";")[0] ?? "";
    const eq = first.indexOf("=");
    if (eq <= 0) continue;
    map.set(first.slice(0, eq).trim(), first.slice(eq + 1).trim());
  }
  return [...map.entries()].map(([k, v]) => `${k}=${v}`).join("; ");
}

/** token 编解码：内部存储 JSON {csrf, cookie} */
function encodeToken(csrf: string, cookie: string): string {
  return Buffer.from(JSON.stringify({ csrf, cookie }), "utf8").toString(
    "base64",
  );
}

function decodeToken(token: string): { csrf: string; cookie: string } {
  let raw: string;
  try {
    raw = Buffer.from(token, "base64").toString("utf8");
  } catch {
    throw new Error("disposablemail-com: token 解码失败");
  }
  const obj = JSON.parse(raw) as { csrf?: string; cookie?: string };
  return { csrf: obj.csrf ?? "", cookie: obj.cookie ?? "" };
}

/**
 * 创建 disposablemail.com 临时邮箱
 * 1. GET / 获取 session cookie 和 CSRF token（CSRF="xxx"）
 * 2. GET /index/index?csrf_token={csrf} 创建邮箱
 * token 以 base64 JSON 存储 {csrf, cookie}。
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  const r1 = await fetchWithTimeout(`${BASE_URL}/`, {
    headers: browserHeaders(),
  });
  if (!r1.ok) {
    throw new Error(`disposablemail-com: 获取首页失败 HTTP ${r1.status}`);
  }
  let cookie = mergeCookieHeader("", r1.headers);
  const html = await r1.text();

  const m = CSRF_RE.exec(html);
  if (!m?.[1]) {
    throw new Error("disposablemail-com: 未能从首页提取 CSRF token");
  }
  const csrf = m[1];

  const r2 = await fetchWithTimeout(
    `${BASE_URL}/index/index?csrf_token=${encodeURIComponent(csrf)}`,
    {
      headers: ajaxHeaders(cookie),
    },
  );
  if (!r2.ok) {
    throw new Error(`disposablemail-com: 创建邮箱失败 HTTP ${r2.status}`);
  }
  cookie = mergeCookieHeader(cookie, r2.headers);

  const data = (await r2.json()) as { email?: string };
  const email = (data.email ?? "").trim();
  if (!email || !email.includes("@")) {
    throw new Error(`disposablemail-com: 获取到的邮箱地址无效: ${email}`);
  }

  return { channel: CHANNEL, email, token: encodeToken(csrf, cookie) };
}

/**
 * 获取 disposablemail.com 邮件列表
 * 1. GET /index/refresh 获取邮件列表（空收件箱返回数字 0）
 * 2. 对每封邮件 GET /email/id/{id} 获取 HTML 正文
 */
export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  const addr = email.trim();
  if (!addr) {
    throw new Error("disposablemail-com: 邮箱地址为空");
  }
  const { cookie } = decodeToken(token);

  const r = await fetchWithTimeout(`${BASE_URL}/index/refresh`, {
    headers: ajaxHeaders(cookie),
  });
  if (!r.ok) {
    throw new Error(`disposablemail-com: 获取邮件列表失败 HTTP ${r.status}`);
  }

  /* 空收件箱返回数字 0 或空数组 */
  const trimmed = (await r.text()).trim();
  if (trimmed === "0" || trimmed === "" || trimmed === "[]") {
    return [];
  }

  let list: DisposablemailRow[];
  try {
    const parsed = JSON.parse(trimmed);
    list = Array.isArray(parsed) ? parsed : [];
  } catch {
    return [];
  }
  if (list.length === 0) {
    return [];
  }

  const emails: Email[] = [];
  for (const item of list) {
    const id = String(item.id ?? "");
    if (!id) {
      continue;
    }
    const htmlBody = await fetchBody(cookie, id);

    const raw: Record<string, unknown> = {
      id,
      from: item.od ?? "",
      to: addr,
      subject: item.predmet ?? "",
      html: htmlBody,
      date: item.kdy ?? "",
      isRead: item.precteno === "precteno",
    };
    emails.push(normalizeEmail(raw, addr));
  }

  return emails;
}

/** 获取单封邮件的 HTML 正文（GET /email/id/{id}） */
async function fetchBody(cookie: string, id: string): Promise<string> {
  if (!id) {
    return "";
  }
  const r = await fetchWithTimeout(`${BASE_URL}/email/id/${id}`, {
    headers: ajaxHeaders(cookie),
  });
  if (!r.ok) {
    return "";
  }
  return await r.text();
}
